packages/authentication/src/types.ts

```diff
@@ -51,3 +51,13 @@ export interface AuthenticationStrategy {
    */
   authenticate(request: Request): Promise<UserProfile | undefined>;
 }
+
+/**
+ * The error thrown when the authentication strategy resolver
+ * cannot find the specified authentication strategy by name.
+ */
+export class AuthenticationStrategyNotFoundError extends Error {
+  constructor(error: string) {
```

Comment:
You may want to add a `code` property as the machine-readable error code that could be understood by any client(HTTP/GRPC/SOAP...)

Reply:
> You may want to add a code property as the machine-readable error code that could be understood by any client(HTTP/GRPC/SOAP...)
@jannyHou , not sure what you mean here.

Reply:
@emonddr The same topic as what we discussed here https://github.com/strongloop/loopback-next/pull/2763#discussion_r278605367 :) 

```ts
export class AuthenticationStrategyNotFoundError extends Error {
  // the `code` would be a string that could be understood by different clients
  // like 'INVALID_ACCESS_TOKEN'
  // see https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/authentication-strategies/JWT.strategy.ts#L35
  code: string;
  constructor(error: string) { 
    super(error);
  }
}
```

Also see my example https://github.com/strongloop/loopback-next/pull/2746/commits/9bda8b7852db196b72129ea6323635b3b46a7d71#diff-440925ebccbc865e10eb2da8810e08eaR9

Reply:
@jannyHou , I removed the `AuthenticationStrategyNotFoundError` class, and now I just create an instance of Error, and augment it (with Object.assign) with a `code` field  which contains a specific error.

```ts
    return this.findAuthenticationStrategy(name).then(strategy => {
      if (strategy) {
        return strategy;
      } else {
        // important not to throw a non-protocol-specific error here
        let error = new Error(`The strategy '${name}' is not available.`);
        Object.assign(error, {
          code: AUTHENTICATION_STRATEGY_NOT_FOUND,
        });
        throw error;
      }
    });

```

Reply:
@emonddr thank you, that refactor looks perfect to me 👍